### PR TITLE
TINY-6001: Improve column resizing behaviour for tables

### PR DESCRIPTION
**Pull request by ltrouton**
Description of Changes:
 - Normalised the column resizing logic so that dragging columns is consistent between fixed and relative table sizes
 - Fixed an issue for relative resizing where the total sum of the column widths did not match the expected table width
 - Added a new `table_column_resizing` setting to allow for different column resizing behaviours

Setting options:
- 'default'
	- dragging an inner column bar resizes only the adjacent column widths
 	- dragging the last column bar resizes the table width whilst preserving ratios between all of the column widths
- 'resizetable'
	- dragging a column bar resizes only the left column width, leaving the other column widths unchanged. As result, the table width resizes.

Pre-checks:
* [X] Changelog entry added
* [X] Tests have been added (if applicable)
* [X] Branch prefixed with `feature/` for new features (if applicable)
* [X] ~~License headers added on new files (if applicable)~~

Review:
* [X] Milestone set
* [x] Review comments resolved

### CODE REVIEW

File: modules/snooker/src/main/ts/ephox/snooker/calc/Deltas.ts [hunk 1]
```diff
@@ -58,7 +63,20 @@ const determine = function (input: number[], column: number, step: number, table
     return onChange(index, next);
   };
 
+  // Applies to the last column bar
   const onRight = function (_prev: number, index: number) {
+    // 'default' means that the whole table and its columns need to be resized to maintain ratio
+    if (columnResizeBehaviour === 'default') {
```
Comment (lnewson) [inline modules/snooker/src/main/ts/ephox/snooker/calc/Deltas.ts:69]: I don't really like all this branching code, as it makes it hard to follow/more complex. It'd be better if we can pass an algorithm in that takes the required values and returns the relevant result (similar to how TableSize works).

File: modules/snooker/src/main/ts/ephox/snooker/resize/TableSize.ts [hunk 2]
```diff
@@ -6,36 +6,40 @@ import * as Sizes from './Sizes';
...
   width: () => number;
   pixelWidth: () => number;
   getWidths: (warehouse: Warehouse, direction: BarPositions<ColInfo>, tableSize: TableSize) => number[];
+  getNewWidths: (widths: number[], deltas: number[]) => number[];
   getCellDelta: (delta: number) => number;
   singleColumnWidth: (w: number, delta: number) => number[];
   minCellWidth: () => number;
   setElementWidth: (cell: Element, amount: number) => void;
   setTableWidth: (table: Element, newWidths: number[], delta: number) => void;
+  widthType: WidthType;
```
Comment (lnewson) [inline modules/snooker/src/main/ts/ephox/snooker/resize/TableSize.ts:21]: This really should not exist for branching purposes. Any branching specific code should be here instead and called.

Comment (lnewson) [reply modules/snooker/src/main/ts/ephox/snooker/resize/TableSize.ts:21]: Also worth noting, if adding new sizing modes, then it means you need to add the branching everywhere you'd done it. So makes adding new feature or future maintenance much harder doing what you've currently done. As an example, these changes wouldn't work with my changes for the new "none" sizing because of this.

File: modules/snooker/src/main/ts/ephox/snooker/calc/Deltas.ts (hunk 1 above)
Comment (metricjs) [reply modules/snooker/src/main/ts/ephox/snooker/calc/Deltas.ts:69]: Mm, I think we've all been talking about potentially doing that in various separate conversations? I'm definitely for it

File: modules/snooker/src/main/ts/ephox/snooker/calc/Deltas.ts [hunk 3]
```diff
@@ -28,7 +29,7 @@ const neighbours = function (input: number[], index: number) {
  * Calculate the offsets to apply to each column width (not the absolute widths themselves)
  * based on a resize at column: column of step: step. The minimum column width allowed is min
```
Comment (metricjs) [inline modules/snooker/src/main/ts/ephox/snooker/calc/Deltas.ts:30]: Is the "min" part of this comment legacy? I assumed it was an arg at first

File: modules/snooker/src/main/ts/ephox/snooker/calc/Deltas.ts (hunk 1 above)
Comment (metricjs) [reply modules/snooker/src/main/ts/ephox/snooker/calc/Deltas.ts:69]: Would tie into the "init like Cement" idea?

File: modules/tinymce/src/plugins/table/main/ts/api/Settings.ts [hunk 4]
```diff
@@ -56,6 +57,12 @@ const getTableHeaderType = (editor: Editor): string => {
   }
 };
 
+const getColumnReszingBehaviour = (editor: Editor): ColumnResizing => {
+  const validModes: ColumnResizing[] = [ 'default', 'static', 'resizetable' ];
+  const givenMode = editor.getParam('table_column_sizing', 'default', 'string');
+  return Arr.find(validModes, (mode) => mode === givenMode).getOr('default');
```
Comment (metricjs) [inline modules/tinymce/src/plugins/table/main/ts/api/Settings.ts:63]: I think I like this way of doing this better than how I've been doing it :D

File: modules/snooker/src/main/ts/ephox/snooker/calc/Deltas.ts (hunk 3 above)
Comment (ltrouton) [reply modules/snooker/src/main/ts/ephox/snooker/calc/Deltas.ts:30]: Yeah, looks like it was originally an arg based on the git history. I'll remove that sentence since it is not relevant anymore.

File: modules/tinymce/changelog.txt [hunk 5]
```diff
@@ -5,6 +5,7 @@ Version 5.4.0 (TBD)
     Added icons for `table` column and row cut, copy and paste toolbar buttons #TINY-6062
     Added all `table` menu items to the UI registry, so they can be used by name in other menus #TINY-4866
     Added new `mceTableApplyCellStyle` command to the `table` plugin #TINY-6004
+    Added new `table_column_sizing` setting to control how the table is affected when resizing columns via the resize bars #TINY-6001
```
Comment (tylerkelly13) [inline modules/tinymce/changelog.txt:8]: please replace 'via' with 'using'

File: modules/tinymce/changelog.txt [hunk 6]
```diff
@@ -19,6 +20,7 @@ Version 5.4.0 (TBD)
     Fixed filters for screening commands from the undo stack to be properly case-insensitive #TINY-5946
     Fixed `fullscreen` plugin now removes all classes when the editor is closed #TINY-4048
     Fixed handling of mixed-case icon names by having the silver theme downcase them before icon retrieval #TINY-3854
+    Fixed an issue where column resizing via the resize bars was inconsistent between fixed and relative table widths #TINY-6001
```
Comment (tylerkelly13) [inline modules/tinymce/changelog.txt:23]: please replace 'via' with 'using'

File: modules/tinymce/changelog.txt (hunk 5 above)
Comment (ltrouton) [reply modules/tinymce/changelog.txt]: Updated

**Comment by ltrouton** [comment]
The code has been updated with a new architecture where the resizing behaviours are defined in ResizeBehaviour.ts and passed around. Huge thanks to @lnewson for helping with this.

One thing that still needs to be looked into is the TinyMCE resizers and corner handles as they currently have their own logic. This will be split it another ticket. Additionally, row resizing will have to be looked into as well at some point.

File: modules/snooker/src/main/ts/ephox/snooker/api/ResizeBehaviour.ts [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+  };
+
+  const calcMiddleDeltas = (input: number[], _prev: number, index: number, next: number, delta: number, minCellSize: number, relativeSizing: boolean) =>
+    calcLeftEdgeDeltas(input, index, next, delta, minCellSize, relativeSizing);
+
+  const resizeTable = (resizer: TableResizer, delta: number) =>
+    resizer(delta);
+
+  // Calculations for the last column/row resizer
+  // Resizing the last column/row only requires the delta of the last column/row to be calcuated
```
Comment (lnewson) [inline modules/snooker/src/main/ts/ephox/snooker/api/ResizeBehaviour.ts:63]: This comment is misleading, as that's true for "fixed" size tables, but for relative tables all rows/cols needs to be adjusted.

File: modules/tinymce/src/plugins/table/main/ts/actions/ResizeHandler.ts [hunk 8]
```diff
@@ -57,9 +57,10 @@ export const getResizeHandler = function (editor: Editor): ResizeHandler {
     const direction = TableDirection(Direction.directionAt);
     const rawWire = TableWire.get(editor);
     wire = Option.some(rawWire);
-    if (hasObjectResizing(editor) && hasTableResizeBars(editor)) {
+    if (Settings.hasObjectResizing(editor) && Settings.hasTableResizeBars(editor)) {
       const lazySizing = (table: SugarElement<HTMLTableElement>) => TableSize.get(editor, table);
-      const sz = TableResize.create(rawWire, direction, lazySizing);
+      const resizing = Settings.getColumnResizingBehaviour(editor) === 'resizetable' ? ResizeBehaviour.resizeTable() : ResizeBehaviour.preserveTable();
```
Comment (lnewson) [inline modules/tinymce/src/plugins/table/main/ts/actions/ResizeHandler.ts:62]: There really should be a TinyMCE test for this part of the logic, to make sure the right mode to selected based on the setting used.

File: modules/snooker/src/main/ts/ephox/snooker/api/ResizeBehaviour.ts (hunk 7 above)
Comment (ltrouton) [reply modules/snooker/src/main/ts/ephox/snooker/api/ResizeBehaviour.ts]: Have adjusted the comments so hopefully they are better now.

File: modules/snooker/src/main/ts/ephox/snooker/api/ResizeBehaviour.ts [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+const zero = (array: number[]) => Arr.map(array, Fun.constant(0));
+
+const surround = (input: number[], startIndex: number, endIndex: number, results: number[], f: (array: number[]) => number[]) =>
+  f(input.slice(0, startIndex)).concat(results).concat(f(input.slice(endIndex)));
+
+// Clamp positive or negative delta so that a column/row cannot be reduced past its min size
+const clampDeltaHelper = (predicate: (delta: number) => boolean) => (sizes: number[], index: number, delta: number, minCellSize: number) => {
+  if (!predicate(delta)) {
+    return delta;
+  }
```
Comment (lnewson) [inline modules/snooker/src/main/ts/ephox/snooker/api/ResizeBehaviour.ts:22]: Please use an `else` here. It produces clearer code and also minifies better.

File: modules/snooker/src/demo/ts/ephox/snooker/demo/DetectDemo.ts [hunk 10]
```diff
@@ -131,9 +132,9 @@ Ready.execute(function () {
   InsertAll.append(ephoxUi, [ ltrs, rtls ]);
 
   const lazyTableSize = (table: Element<HTMLTableElement>) => TableSize.getTableSize(table);
-  const ltrManager = TableResize.create(ResizeWire.body(tester, ltrs), ResizeDirection.ltr, lazyTableSize);
+  const ltrManager = TableResize.create(ResizeWire.body(tester, ltrs), ResizeDirection.ltr, ResizeBehaviour.preserveTable(), lazyTableSize);
   ltrManager.on();
-  const rtlManager = TableResize.create(ResizeWire.body(subject3, rtls), ResizeDirection.rtl, lazyTableSize);
+  const rtlManager = TableResize.create(ResizeWire.body(subject3, rtls), ResizeDirection.rtl, ResizeBehaviour.resizeTable(), lazyTableSize);
```
Comment (lnewson) [inline modules/snooker/src/demo/ts/ephox/snooker/demo/DetectDemo.ts:137]: This was probably my fault, but this should probably be `preserveTable` here.

File: modules/snooker/src/main/ts/ephox/snooker/api/ResizeBehaviour.ts [hunk 11]
```diff
@@ -0,0 +1,144 @@
+import { Arr, Fun } from '@ephox/katamari';
+
+type TableResizer = (delta: number) => void;
+
+export interface ResizeBehaviour {
+  readonly resizeTable: (resizer: TableResizer, delta: number, isLastColumn: boolean) => void;
+  readonly clampTableDelta: (sizes: number[], index: number, delta: number, minCellSize: number, isLastColumn: boolean) => number;
+  readonly calcLeftEdgeDeltas: (input: number[], index: number, nextIndex: number, delta: number, minCellSize: number, relativeSizing: boolean) => number[];
```
Comment (lnewson) [inline modules/snooker/src/main/ts/ephox/snooker/api/ResizeBehaviour.ts:8]: Oops I forgot to rename `input` here. If you get a chance, mind renaming these to `sizes`?

File: modules/snooker/src/main/ts/ephox/snooker/api/ResizeBehaviour.ts (hunk 9 above)
Comment (ltrouton) [reply modules/snooker/src/main/ts/ephox/snooker/api/ResizeBehaviour.ts:26]: Fixed

File: modules/snooker/src/demo/ts/ephox/snooker/demo/DetectDemo.ts (hunk 10 above)
Comment (ltrouton) [reply modules/snooker/src/demo/ts/ephox/snooker/demo/DetectDemo.ts]: Fixed

File: modules/snooker/src/main/ts/ephox/snooker/api/ResizeBehaviour.ts (hunk 11 above)
Comment (ltrouton) [reply modules/snooker/src/main/ts/ephox/snooker/api/ResizeBehaviour.ts]: Done

Comment (lnewson) [reply modules/snooker/src/main/ts/ephox/snooker/api/ResizeBehaviour.ts]: Thanks for that :+1:

File: modules/snooker/src/main/ts/ephox/snooker/api/Main.ts [hunk 12]
```diff
@@ -32,6 +33,7 @@ export {
   CopyRows,
   CopySelected,
   Generators,
+  ResizeBehaviour,
```
[the remark this answers is not in the document]
Comment (lnewson) [reply modules/snooker/src/main/ts/ephox/snooker/api/Main.ts:36]: Good catch, we'll need a version bump added in `versions.txt` so we know to bump it next time we release.

- [ ] Add snooker version to `versions.txt`
